docs/proposals/11-apiserver-network-proxy.md

```diff
@@ -0,0 +1,108 @@
+
+`APIServer <-> Proxy-Server <--- internet <--- Proxy-Agent --> Pods | Nodes | Services`
+
+
+### API Server Network Proxy
+
+To solve this issue we can utilize the [apiserver-network-proxy](apiserver-network-proxy) upstream implementation. Which provides a reference implementation for a reverse streaming server. The way it works is as follows:
+
+- Proxy agent connects to proxy server to establish a sticky connection.
+- Traffic to the proxy server gets then re-directed to the agent which forwards the traffic to in-cluster components.
+
+The initial motivation for the apiserver-network-proxy project is to get rid of provider-specific implementations that reside in the API-server (e.g., SSH), but it turns out that
+it has other interesting use-cases such as data-plane connection decoupling, which is the main use-case for this proposal.
+
+Starting **Kubernetes 1.18** its possible to make use of an ```--egress-selector-config-file``` flag, this helps point the API-server to traffic hook points based on traffic direction, for example, in the config below, the API server would have to forward all cluster related traffic (e.g., logs, port-forward, exec, ...etc.) to the **proxy-server** which then knows how to forward traffic to the shoot. For the rest of the traffic e.g., API server to ETCD or other control-plane components `direct` is used which means legacy routing method, i.e., by-pass the proxy.
+
+```yaml
+  egress-selector-configuration.yaml: |-
+    apiVersion: apiserver.k8s.io/v1alpha1
+    kind: EgressSelectorConfiguration
+    egressSelections:
+    - name: cluster
+      connection:
+        proxyProtocol: httpConnect
+        transport:
+          tcp:
+            url: https://proxy-server:8131
+    - name: master
+      connection:
+        proxyProtocol: direct
+    - name: etcd
+      connection:
+        proxyProtocol: direct
+```
+
+### Challenges
+
+One challenge remains to completely eliminate the need for a VPN connection. In today's Gardener setup, each control-plane has a Prometheus instance that directly scrapes cluster components such as CoreDNS, Kubelets, cadvisor, etc. This works because in addition to the VPN side car attached to the API server pod, we have another one attached to prometheus which knows how to forward traffic to these endpoints. Once the VPN is eliminated, it is required to find other means to forward traffic to these components.
+
+
+### Solutions
+
+There are currently two ways to solve this problem:
+
+- Attach a port-forwarder side-car to prometheus.
+- Utilize the proxy subresource on the API server.
+
+#### Port-forwarder Sidecar
+
+With this solution each prometheus instance would have a side-car that has the kubeconfig of the shoot cluster, and which estalishes a port-forward connection to the endpoints residing in the shoot.
+
+There are a couple of problems with this approach:
+
+- the port-forward connection is not reliable.
+- the connection would break if the API server instance dies.
```

Comment:
OK, however, as far as I have understood our control plane auto-scaling, which is the main source for dying API servers, that's handled in most cases gracefully (OOM isn't of course, but shouldn't happen often).

But of course you are right. We don't want/need false positive alerts. Weighting it, I love getting rid of the VPN LB for the reasons stated above and we could maybe tune the monitoring alerts (wait longer/for more missed probes, which should reduce the likelihood of false positives, right?).